### ISSUE TITLE
We should document that PropertyT is like a non-deterministic monad

### ISSUE THREAD

**Issue by ocharles** [hedgehogqa/haskell-hedgehog#122]
I was just playing around with interleaving some database calls in a `PropertyT` computation, and confused myself. I've now come to realise that `PropertyT` is like a non-deterministic monad (e.g., list, `LogicT`) and that performing IO doesn't necessarily imply it will all be performed in exactly the same order over shrinks. For example,

```haskell
property $ do
  foo <- forAll $ pure True
  liftIO $ putStrLn "blah"
  bar <- forAll $ Gen.integral (Range.linear 1 100)
  ...
```

Will only ever execute the `putStrLn` once, as everything "above" it is a one-node `GenT` tree (as confirmed by `pure` with `forAll`). If something fails on the value of `bar`, the next thing to do is shrink `bar`, and that is where execution will resume from.

If I'm correct in my understanding, I think it would be good to write this down.

**Comment by jystic** [comment]
Yeah that's correct, it's always been a concern of mine that allowing effects like this is a bit confusing, but I didn't want to rule it out in case someone found a good use case for it.

**Comment by jystic** [comment]
I think the example in the above post could be made to work using `ReaderT`, something like:

```haskell
s_coin :: (
    Monad gen
  , MonadIO m
  , MonadTest m
  , MonadReader Turnstile m
  )
  => Turnstile
  -> Command gen m ModelState
s_coin =
  ...

prop_turnstile :: Property
prop_turnstile =
  property $ do
    actions <- forAll $
      Gen.sequential (Range.linear 1 100) initialState [
          s_coin
        , s_push_locked
        , s_push_unlocked
        ]
    turnstile <- liftIO newTurnstile
    flip runReaderT turnstile $ executeSequential initialState actions
```

**Comment by massysett** [comment]
Hi Jacob,

Thanks for all your work on Hedgehog.  It's a huge advance in the realm of property testing, and it has saved me a lot of work and frustration.  The elegance of the library and the thought that went into its API is impressive.

I agree that it would be nice to document that `PropertyT` can be non-deterministic.  This bit me when I tried the state-machine testing.  I'm testing something that uses SQLite, and I figured I'd be a good citizen by closing my SQLite database handles after use.  I had a test that failed (as expected) but the failure showed that it arose from misuse of the SQLite library (it literally will report `SQLITE_MISUSE`.).  For awhile I wondered if there was a bug in my SQLite Haskell bindings, but then realized that `SQLITE_MISUSE` arose only during shrinking - without shrinking, the test failed for the expected reason (a constraint violation).  It seems that what's happening is that on first run the handle will be closed, but during shrinking the handle is used again, so SQLite complains that I'm using a closed handle.  Using [newer methods](https://jacobstanley.io/5-tips-for-better-hedgehog-tests/) to manage resources (instead of the [old, ResourceT methods](https://github.com/hedgehogqa/haskell-hedgehog/blob/master/hedgehog-example/src/Test/Example/Resource.hs)) didn't help.The problem went away when I simply stopped cleaning up my database handles.

I've written around this problem by using entirely pure computations, which is quite doable since SQLite can run entirely in memory.  For the time being this has scared me away from ever doing any IO in Hedgehog properties because I simply do not understand how it works.  I'd recommend removing the MonadIO instance altogether since its behavior is so surprising or, barring that, attaching a big warning that doing IO in properties can yield bizarre results.
